whole-tale/girder_wholetale
# Unified notification support

See the [Notifications design doc](https://docs.google.com/document/d/1IkvsZqxxlIBaxVLT1yb1kWtGyDuJRo8ZzT1hsqke_NE) for details.

Along with https://github.com/whole-tale/gwvolman/pull/75, this PR implements changes to how we handle task notifications and progress reporting for instance creation, restarting, and image building.
* Each logical task (which may be multi-job) creates a progress notification and passes the ID to each job via `girder_job_other_fields`. This results in the job object having a `wt_notification_id` field.
* Gwvolman (Celery) tasks use the internal `job_manager.updateProgress()` method to update per-job progress.
* A generic `updateNotification` event handler (event=jobs.job.update.after) updates a single progress notification based on the provided `wt_notification_id`. This allows us to accumulate progress across multiple jobs and control the notification payload.
* The `resource` field  (i.e., payload) for each notification contains relevant information including a `type` (e.g., wt_build_image) and identifiers of objects related to the notification (e.g., tale ID, job IDs)
* For now, progress notifications have a default expiration of 10 days.  This is intended to allow us to support the "Archived notifications" feature in the UI.

To test (assumes `deploy-dev`):
* Checkout `notification-poc` branches for deploy-dev, girder_wholetale and gwvolman
* Restart Girder and gwvolman as needed
* In deploy-dev, run `test_stream.py <valid girder token>` (which requires the sseclient package). This will listen to the notification stream from now.
* Via the UI or swagger, launch an instance.  You should see 7 notifications in the stream -- 1 for each step (each job has two updates: starting, complete) and a final success status change.
* Restart and rebuild should each have 3 notifications including the success status change.
